lib/ptime/syndic_date.ml

```diff
@@ -0,0 +1,163 @@
+      else
+        (* FIXME: this should be made more robust. *)
+        let tz_offset_s =
+          match z with
+          | "EST" -> (-5) * 3600
+          | "EDT" -> (-4) * 3600
+          | "CST" -> (-6) * 3600
+          | "CDT" -> (-5) * 3600
+          | "MST" -> (-7) * 3600
+          | "MDT" -> (-6) * 3600
+          | "PST" -> (-8) * 3600
+          | "PDT" -> (-7) * 3600
+          | _ ->
+            let zh = sscanf (String.sub z 0 3) "%i" (fun i -> i) in
+            let zm = sscanf (String.sub z 3 2) "%i" (fun i -> i) in
+            let tz_sign = if zh < 0 then -1 else 1 in
+            if zh < 0 then
+              tz_sign * (((-zh) * 3600) + (zm * 60))
+            else tz_sign * ((zh * 3600) + (zm * 60))
+        in
+        let rt = map2 (fun date span -> Ptime.add_span date span) date span in
+        (* XXX: We lose minutes with this conversion, but Calendar does not
+           propose to handle minutes. *)
+        map (fun x -> Some (Ptime.to_date_time ~tz_offset_s x)) rt
+      in match map Ptime.of_date_time date_and_time with
+      | Some x -> x
+      | None -> invalid_arg(sprintf "Syndic.Date.of_rfc822: cannot parse")
+  in
+  try
+    if 'A' <= s.[0] && s.[0] <= 'Z' then (
+      try sscanf s "%_s %i %s %i %i:%i%s %s" make_date
+      with _ -> sscanf s "%_s %ist %s %i %i:%i%s %s" make_date
+    )
+    else (
+      try sscanf s "%i %s %i %i:%i%s %s" make_date
+      with _ -> sscanf s "%i %s %i" (fun d m y -> make_date d m y 0 0 "" "UT")
+    )
+  with _ ->
+    invalid_arg(sprintf "Syndic.Date.of_string: cannot parse %S" s)
+
+type month =
+  Jan | Feb | Mar | Apr | May | Jun | Jul | Aug | Sep | Oct | Nov | Dec
+
+let string_of_month = function
+  | Jan -> "Jan"  | Feb -> "Feb"  | Mar -> "Mar"  | Apr -> "Apr"  | May -> "May"
+  | Jun -> "Jun"  | Jul -> "Jul"  | Aug -> "Aug"  | Sep -> "Sep"  | Oct -> "Oct"
+  | Nov -> "Nov"  | Dec -> "Dec"
+
+let month_of_date =
+  let months = [| Jan; Feb; Mar; Apr; May; Jun; Jul; Aug; Sep; Oct; Nov; Dec |] in
+  fun t ->
+  let (_, i, _) = Ptime.to_date t in
+  months.(i)
+
+(* RFC3339 date *)
+let of_rfc3339 s = Ptime.of_rfc3339 s
+  |> function Result.Ok (t, Some tz_offset_s, _) ->
+              Ptime.to_date_time ~tz_offset_s t
+              |> Ptime.of_date_time
+              |> (function Some x -> x
+                         | None -> invalid_arg(sprintf "Syndic.Date.of_string: cannot parse %S" s))
+            | Result.Ok (t, None, _) ->
+              Ptime.to_date_time t |> Ptime.of_date_time
+              |> (function Some x -> x
+                         | None -> invalid_arg(sprintf "Syndic.Date.of_string: cannot parse %S" s))
+            | _ -> invalid_arg(sprintf "Syndic.Date.of_string: cannot parse %S" s)
+
+let to_rfc3339 d =
+  (* Example: 2014-03-19T15:51:25.050-07:00 *)
+  Ptime.to_rfc3339 d
+
+(* Convenience functions *)
+
+let day_of_week =
+  let wday = [| `Thu; `Fri; `Sat; `Sun; `Mon; `Tue; `Wed |] in
```

Comment:
It is odd to use polymorphic variants here if standard variants are used for months.